src/transformers/models/flava/modeling_flava.py

```diff
@@ -0,0 +1,2101 @@
+            # Note that these states don't use final layernorm in the transformer model
+            text_mm_projection = self.text_to_mm_projection(text_states[-1])
+
+        multimodal_embeddings = None
+        multimodal_output = None
+        if image_mm_projection is not None and text_mm_projection is not None and not skip_multimodal_encoder:
+            multimodal_input = torch.cat([image_mm_projection, text_mm_projection], dim=1)
+            multimodal_output = self.multimodal_model(multimodal_input, return_dict=return_dict)
+            multimodal_embeddings = multimodal_output[0]
+
+        if not return_dict:
+            return (
+                image_embeddings,
+                image_output,
+                text_embeddings,
+                text_output,
+                multimodal_embeddings,
+                multimodal_output,
+            )
+
+        return FLAVAModelOutput(
+            image_embeddings=image_embeddings,
+            image_output=image_output,
+            text_embeddings=text_embeddings,
+            text_output=text_output,
+            multimodal_embeddings=multimodal_embeddings,
+            multimodal_output=multimodal_output,
+        )
+
+
+class FLAVACodebookBlock(nn.Module):
+    def __init__(self, in_size: int, out_size: int, num_layers: int, **kwargs):
+        super().__init__()
+
+        n_hid = out_size // 4
+        self.post_gain = 1 / (num_layers**2)
+
+        if in_size != out_size:
+            self.id_path = _build_codebook_conv2d(in_size, out_size, kernel_size=1)
+        else:
+            self.id_path = nn.Identity()
+
+        self.res_path = nn.Sequential(
+            OrderedDict(
+                [
+                    ("relu_1", nn.ReLU()),
+                    ("conv_1", _build_codebook_conv2d(in_size, n_hid, kernel_size=3)),
+                    ("relu_2", nn.ReLU()),
+                    ("conv_2", _build_codebook_conv2d(n_hid, n_hid, kernel_size=3)),
+                    ("relu_3", nn.ReLU()),
+                    ("conv_3", _build_codebook_conv2d(n_hid, n_hid, kernel_size=3)),
+                    ("relu_4", nn.ReLU()),
+                    ("conv_4", _build_codebook_conv2d(n_hid, out_size, kernel_size=1)),
+                ]
+            )
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.id_path(x) + self.post_gain * self.res_path(x)
+
+
+# Inspired by DALLE Encoder in https://github.com/openai/DALL-E/blob/5be4b236bc3ade6943662354117a0e83752cc322/dall_e/encoder.py#L42
+@add_start_docstrings(
+    """
+    The FLAVA's codebook model inspired from DALL-E's original encoder. Outputs raw hidden states and can be used to
+    generate image tokens for an image based on DALL-E's vocab. To be used to generate labels for MIM. Use
+    `get_codebook_indices` to get image tokens for an image.
+    """,
+    FLAVA_START_DOCSTRING.format(config="FLAVACodebookConfig"),
+)
+class FLAVACodebook(FLAVAPreTrainedModel):
```

Comment:
Just to understand this a bit better - would this model class every be used as a stand-alone model or in inference? Or does it purely exist for pretraining? 

If this model is only uses for pretraining (MIM), I'd prefer to not make it its own class, but to just add it as a submodule of `...ForPreTraining`. It's quite important that we only have one modeling file IMO and don't create dependencies between different model repos that can get out of sync.

The way I currently understand it is that this model class should never be used as a standalone model

Reply:
There are multiple reasons:
1. Codebook generates mim labels only which are not needed for running the inference on the model. It will unnecessarily bloat the model if we added codebook to the pretrained model even if the user doesn't need. As others have discussed, logic for generating `mim_labels` should be independent of the model itself.
2. There can be any other codebook used with flava not just this one. Keeping it separate helps with that.
3. FLAVACodebook can also be used independently of FLAVA for other projects. It is very similar to DALLE's Encoder.

Reply:
Ok but then let's just add `codebook` to `ForPreTraining` but not to `Model` then it would not bloat inference no?

Reply:
same for `ForImageClassification` - no need to add it there, but think it's better to add it to `ForPretraining` so that a single forward pass of `ForPretraining` is enough to get a loss

Reply:
A lot of tasks such as WinoGround, zero-shot classification, ITM retrieval use `ForPretraining` out-of-the-box and that's what I meant by inference. I would really like to avoid generating MIM labels unnecessarily inside the model as they are not even needed for inference. This is in general the case for `targets` no? Models are provided targets they don't generate them, unless it is a teacher-student model but this case it is not. Wdyt?

Reply:
For me it's weird / wrong if a model class that is called `ForPretraining` is used for zero-shot classification in inference, but don't really know WinoGround and ITM retrieval - but think we could give the classes also different names in this case (happy with `ForPretraining` just fro pretraining and other names just for inference that don't need a codebook). 

I don't think we should generate MLM lables inside the model (we should never do this - labels should be passed to the model). For inference I don't think we ever need to call the codebook. I would be in favor of adding a input arg `labels` or `masked_image_indices` to the forward method of `ForPretraning` and only if those are provided then the codebook is called - otherwise we don't call the codebook. It could be done in a very similar way to how it's done for Wav2Vec2 IMO. E.g. the quantizer is part of the `ForPretraining` model but only used if `sampled_negativ_indices` is used - see: https://github.com/huggingface/transformers/blob/67ed0e43dcdcab742e4552cabe04f88e97b5766b/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L1537 

If we also really want to optimize for not requiring parameters unnecessary, but still need one `ForPretraining` class for both inference and training we could do something like:


```python
if config.add_codebook:
    self.codebook = ...
else:
    self.codebook = None
``` 

in the init of `ForPretraining`

Reply:
Just to clarify, since I think we are on the same page, `codebook` is the one that generates `mim_labels`. Also, I am talking `mim_labels` not `mlm_labels`.

Reply:
I am not following your argument Patrick. You seem to want contradictory things at the same time:
- the model should not generate its labels, they should be passed to the model (I 100% agree)
- the codebook should be added inside the model for pretraining, yet it would only be there to generate the mim labels, so it would contradict 1.

Reply:
Ok I might not have been very clear here.

For 2) the codebook should not generate the labels, it should process the image to the labels (the codebook is an encoder if I understand correctly), which means that by passing a mask that states which part of the codebook output is masked and `pixel_values` or whatever will be processed by the codebook, we get the labels

Reply:
That's what is done for Wav2Vec2

Reply:
For me having a single pytorch file / single model repo is important which is what I'm after here. 

I don't think we have any architectures that require multiple model repos / model directories for inference, fine-tuning , or pretraining (no?). I'd like to keep it that way. RAG, CLIP are all composite models, but they are self-contained in a single model repo. 

cc @patil-suraj @LysandreJik as well here

Reply:
Copy pasting some offline discussion that I had with Patrick for completeness:

P:
> Yeah I think, we're more or less on the same page. Also think we should try to avoid unnecessary using memory and labels should not be created inside the model. But if the creation of labels depends on pretrained weights (which seems to be the case here and which is also the case for Wav2Vec2) IMO we should add the weights to the model architecture and pass something like mim_label_mask_indices that will be the indices that are used to mask the output of codebook
> This has the following advantages:
> - Label creation is deterministic because it all depends on mim_label_mask_indices 
> - Single pytorch file for training (that's very important for us). Otherwise people don't have a self-contained model folder that contains all the weights when doing model.save_pretrained("...") and it creates difficult dependencies between repos
> - We don't really try to have a modularized architecture style, rather a self-contained style 

A:
> I think that is bool_masked_pos.
> What you are saying is correct. Codebook generates `mim_labels` but they are masked using `bool_masked_pos` to understand which one we need to predict.

P:
> Cool couldn't we have `bool_masked_pos` be an input to ForPretraining ? Just like in Wav2Vec2?

A: 
> We have it as of now. mim_labels are masked using this later on. (before Cross Entropy calculation)

> For the moment, I have been using same pretrained weights for FlavaModel and FlavaForPretraining. Using the FlavaForPretraining weights to load the FlavaModel will throw a lot of unexpected keys error but that it fine I guess?

P: 
> This is expected - that's happens everytime I call Wav2Vec2Model.from_pretrained(...)
> Also happens everytime you do BertModel.from_pretrained("bert-base-cased")

A: 
>  Sounds good. I think I am okay with this. I will also add an option `add_codebook` which is true by default then?
> I would also allow user to pass `mim_labels` to the forward. If they do, we use those otherwise we use codebook to generate the labels and mask them.
> Only thing that is left is how do we generate mlm_labels more elegantly.
> The DataCollator doesn't work for me as it just updates input_ids which we need as it is. Maybe a new data collator, I am not sure. There should be something simpler.

P:
> That's what I do for Wav2Vec2: https://github.com/huggingface/transformers/blob/67ed0e43dcdcab742e4552cabe04f88e97b5766b/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L313
> Then this: https://github.com/huggingface/transformers/blob/67ed0e43dcdcab742e4552cabe04f88e97b5766b/examples/pytorch/speech-pretraining/run_wav2vec2_pretraining_no_trainer.py#L326 for pretraining

A:
> I also remembered that the feature extractor will also need to now output separate processed image for code book and forward needs to take it as a parameter. That was one more reason to keep them separate.

P:
> Ah interesting, so you need 1 tokenizer and 2 feature extractors for flava?
> IMO, the best approach would then be to add a `return_pixel_values_for_codebook` parameter to the feature extractor 
> Actually it'd be good if we would discuss this public on GitHub so that Sylvain, Lysandre and Suraj could also take a look and give their opinion 🙂

so after that we are back to GitHub.

Would be good to have your opinions @sgugger @LysandreJik @patil-suraj

Reply:
I agree that everything should be in one repo, we never use two separate repos for one model, so that does imply the codebook should be part of the model that uses it, to be in the same checkpoint (and if the same codebook weights can be reused for different FLAVA models, it's completely supported if the same codebook weights are in all the checkpoints).

The only alternatives (while keeping a single repo) would be to have the weights of the codebook with some adhoc prefix corresponding to the base prefix of the codebook and different than the base prefix of the other FLAVA models, so rather a ugly hack; or use a different branch for the codebook, which would result in `codebook = Xxx.from_pretrainec(checkpoint, revision="codebook" )`.